### PR TITLE
RAP-4131 RAP-4132 RAP-4116 RAP-4137

### PR DESCRIPTION
**Pull request by scarabus**

### CODE REVIEW

File: Libs/RaptureAPI/src/main/java/rapture/common/client/BaseHttpApi.java [hunk 1]
```diff
@@ -250,6 +251,16 @@ public long getContentLength() {
     }
 
     protected GeneralResponse responseFromJson(String responseObjectJson) {
+        // if the servlet throws an exception - which happens if a Reflex assertion fails - then extract the info
```
Comment (jonathan-major) [inline Libs/RaptureAPI/src/main/java/rapture/common/client/BaseHttpApi.java]: please log a ticket for this change. cheers

Comment (jonathan-major) [inline Libs/RaptureAPI/src/main/java/rapture/common/client/BaseHttpApi.java]: @scarabus can you add a ticket for the change to BaseHttpApi class please?

Comment (scarabus) [inline Libs/RaptureAPI/src/main/java/rapture/common/client/BaseHttpApi.java]: Ticket RAP-4137 added

File: Libs/RaptureAPI/src/main/java/rapture/common/client/BaseHttpApi.java [hunk 2]
```diff
@@ -250,6 +251,16 @@ public long getContentLength() {
...
+        if (responseObjectJson.startsWith("<!DOCTYPE html>")) {
+            log.warn(responseObjectJson);
+            int rootCause = responseObjectJson.indexOf("root cause");
+            if (rootCause > 0) {
+                int skipPre = responseObjectJson.indexOf("<pre>", rootCause);
+                String cause = responseObjectJson.substring(skipPre + 5, responseObjectJson.indexOf("</pre>", skipPre)).replaceAll("&quot;", "\"");
+                throw new RaptureException("", 500, cause);
```
Comment (jonathan-major) [inline Libs/RaptureAPI/src/main/java/rapture/common/client/BaseHttpApi.java]: will the response code always be 500?

Comment (scarabus) [inline Libs/RaptureAPI/src/main/java/rapture/common/client/BaseHttpApi.java]: 500 seems the most suitable for the general case. I wrote this specifically to catch the case where a Reflex test case assertion fails and to extract the error message. In such a case we currently get a 500 (which is embedded in the text). I considered extracting it in the same way.

Another approach might be to check the cause in more detail return a specific HTTP code to cover the case of an assertion fail and save 500 for genuine internal server errors. I'm tempted to use 418 or an unused 5xx code.

**Pull request merged**